modalc-lib/modal-vector.rkt

```diff
@@ -0,0 +1,152 @@
+#lang racket
+
+(require modalc
+         syntax/parse/define)
+
+;; apply mode to both ref and set
+(define (modal-vector/c should-apply-ctc? . inner-ctc)
+  ;; passes inner-ctc in as a list
+  (modal-vector/c* inner-ctc #:ref/set should-apply-ctc?))
+
+
+(define (modal-vector/c* #:ref [ref-ctc? #f]
+                         #:set [set-ctc? #f]
+                         #:ref/set [all-ctc? #f]
+                         . inner-ctc)
+  
+  ;; when using modal-vector/c
+  (when (list? (first inner-ctc))
+    (set! inner-ctc (first inner-ctc)))
+
+  ;; error when no mode exists
+  (unless (or ref-ctc?
+              set-ctc?
+              all-ctc?)
+    (raise-user-error 'modal-vector/c* "No mode supplied"))
+  
+  ;; create ref and set modes
+  (define should-apply-ctc?/ref (or ref-ctc?
+                                    all-ctc?
+                                    mode:never))
+  (define should-apply-ctc?/set (or set-ctc?
+                                    all-ctc?
+                                    mode:never))
+  (define inner-ctc-proj
+    (map contract-late-neg-projection inner-ctc))
+  (make-contract
+   #:name `(modal-vector/c ,(map contract-name inner-ctc-proj))
+   #:late-neg-projection
+   (λ (blame)
+     (define inner-ctc-proj/blame (apply vector-immutable (map (λ (p) (p (blame-swap blame))) inner-ctc-proj)))
+     (λ (val neg-party)
+       (impersonate-vector val
+                           (λ (vec index value)
+                             (cond [(should-apply-ctc?/ref (list index value))
+                                    ((vector-ref inner-ctc-proj/blame index) value neg-party)]
+                                   [else value]))
+                           (λ (vec index value)
+                             (cond [(should-apply-ctc?/set vec)
```

Comment:
Why doesn't the `set` modality get the index and value like the `ref` one?

Reply:
It should. I just forgot to implement it in both.